x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/retriever/PinnedRetrieverBuilder.java

```diff
@@ -0,0 +1,245 @@
+
+    private void validateIdsAndDocs(List<String> ids, List<SpecifiedDocument> docs) {
+        if ((ids != null && ids.isEmpty() == false) && (docs != null && docs.isEmpty() == false)) {
+            throw new IllegalArgumentException("Both 'ids' and 'docs' cannot be specified at the same time");
+        }
+    }
+
+    private void validateSort(SearchSourceBuilder source) {
+        List<SortBuilder<?>> sorts = source.sorts();
+        if (sorts == null || sorts.isEmpty()) {
+            return;
+        }
+        for (SortBuilder<?> sort : sorts) {
+            if (sort instanceof ScoreSortBuilder) {
+                continue;
+            }
+            if (sort instanceof FieldSortBuilder) {
+                FieldSortBuilder fieldSort = (FieldSortBuilder) sort;
+                if (ShardDocSortField.NAME.equals(fieldSort.getFieldName())) {
+                    continue;
+                }
+            }
+            throw new IllegalArgumentException("Pinned retriever only supports sorting by score. Custom sorting is not allowed.");
+        }
+    }
+
+    public PinnedRetrieverBuilder(List<String> ids, List<SpecifiedDocument> docs, RetrieverBuilder retrieverBuilder, int rankWindowSize) {
+        super(new ArrayList<>(), rankWindowSize);
+        validateIdsAndDocs(ids, docs);
+        this.ids = ids != null ? ids : new ArrayList<>();
+        this.docs = docs != null ? docs : new ArrayList<>();
+        addChild(new PinnedRetrieverBuilderWrapper(retrieverBuilder));
+    }
+
+    public PinnedRetrieverBuilder(
+        List<String> ids,
+        List<SpecifiedDocument> docs,
+        List<RetrieverSource> retrieverSource,
+        int rankWindowSize,
+        String retrieverName,
+        List<QueryBuilder> preFilterQueryBuilders
+    ) {
+        super(retrieverSource, rankWindowSize);
+        validateIdsAndDocs(ids, docs);
+        this.ids = ids;
+        this.docs = docs;
+        this.retrieverName = retrieverName;
+        this.preFilterQueryBuilders = preFilterQueryBuilders;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    public int rankWindowSize() {
+        return rankWindowSize;
+    }
+
+    /**
+     * Creates a PinnedQueryBuilder with the appropriate pinned documents.
+     * Prioritizes docs over ids if both are present.
+     *
+     * @param baseQuery the base query to pin documents to
+     * @return a PinnedQueryBuilder or the original query if no pinned documents
+     */
+    private QueryBuilder createPinnedQuery(QueryBuilder baseQuery) {
+        if (baseQuery == null) {
+            baseQuery = new MatchAllQueryBuilder();
+        }
+        if (docs.isEmpty() == false) {
+            return new PinnedQueryBuilder(baseQuery, docs.toArray(new SpecifiedDocument[0]));
+        } else if (ids.isEmpty() == false) {
+            return new PinnedQueryBuilder(baseQuery, ids.toArray(new String[0]));
+        } else {
+            return baseQuery;
```

Comment:
This state should return an error, because no docs or ids were specified. Please validate this in the constructor so there's no need to handle this case.